### PR TITLE
Implement LDAP Certificate

### PR DESCRIPTION
**Pull request by sbishel**
#### Summary
Implements uploading and removing Private and public certicates for LDAP authentication.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-27352

#### Related PR
https://github.com/mattermost/mattermost-webapp/pull/6345
https://github.com/mattermost/mattermost-redux/pull/1224
https://github.com/mattermost/enterprise/pull/752
https://github.com/mattermost/mattermost-api-reference/pull/570

### CODE REVIEW

File: api4/ldap_test.go [hunk 1]
```diff
@@ -112,3 +198,43 @@ func TestMigrateIdLdap(t *testing.T) {
...
+	defer th.TearDown()
+
+	// sth.th.App.UpdateConfig(func(cfg *model.Config) { *cfg.SamlSettings.IdpCertificateFile = app.SamlIdpCertificateName })
+	// defer sth.th.App.UpdateConfig(func(cfg *model.Config) { *cfg.SamlSettings.IdpCertificateFile = "" })
+
+	_, resp := th.Client.UploadSamlPrivateCertificate([]byte(spPrivateKey), "private.key")
+	require.NotNil(t, resp.Error, "Should have failed. No System Admin privileges")
+
+	_, resp = th.SystemAdminClient.UploadSamlPrivateCertificate([]byte(spPrivateKey), "private.key")
+	require.Nil(t, resp.Error, "Should have passed. System Admin privileges %v", resp.Error)
```
Comment (isacikgoz) [inline api4/ldap_test.go:233]: Shall we include local client to the tests? (this question applies to other test casess 😄 )
```suggestion
	th.TestForSystemAdminAndLocal(t, func(t *testing.T, client *model.Client4) {
		_, resp := client.UploadSamlPrivateCertificate([]byte(spPrivateKey), "private.key")
		require.Nil(t, resp.Error, "Should have passed. System Admin privileges %v", resp.Error)
	}
```

File: app/ldap.go [hunk 2]
```diff
@@ -175,3 +182,101 @@ func (a *App) MigrateIdLDAP(toAttribute string) *model.AppError {
...
+}
+
+func (a *App) RemoveLdapPublicCertificate() *model.AppError {
+	if err := a.removeLdapFile(*a.Config().LdapSettings.PublicCertificateFile); err != nil {
+		return err
+	}
+
+	cfg := a.Config().Clone()
+	*cfg.LdapSettings.PublicCertificateFile = ""
+	// *cfg.LdapSettings.Encrypt = false
```
Comment (isacikgoz) [inline app/ldap.go:255]: Shall we remove the commented out code? Or add a line for the reason why they are commented out.

File: api4/ldap_test.go [hunk 3]
```diff
@@ -112,3 +198,43 @@ func TestMigrateIdLdap(t *testing.T) {
 		CheckNotImplementedStatus(t, resp)
 	})
 }
+
+func TestUploadPublicCertificate(t *testing.T) {
+	th := Setup(t)
+	defer th.TearDown()
+
+	// sth.th.App.UpdateConfig(func(cfg *model.Config) { *cfg.SamlSettings.IdpCertificateFile = app.SamlIdpCertificateName })
+	// defer sth.th.App.UpdateConfig(func(cfg *model.Config) { *cfg.SamlSettings.IdpCertificateFile = "" })
```
Comment (hahmadia) [inline api4/ldap_test.go:207]: Are we able to delete these commented code or do we need it?

File: api4/ldap_test.go [hunk 4]
```diff
@@ -112,3 +198,43 @@ func TestMigrateIdLdap(t *testing.T) {
...
+	_, resp = th.SystemAdminClient.DeleteLdapPublicCertificate()
+	require.Nil(t, resp.Error, "Should have passed. System Admin privileges %v", resp.Error)
+}
+
+func TestUploadPrivateCertificate(t *testing.T) {
+	th := Setup(t)
+	defer th.TearDown()
+
+	// sth.th.App.UpdateConfig(func(cfg *model.Config) { *cfg.SamlSettings.IdpCertificateFile = app.SamlIdpCertificateName })
+	// defer sth.th.App.UpdateConfig(func(cfg *model.Config) { *cfg.SamlSettings.IdpCertificateFile = "" })
```
Comment (hahmadia) [inline api4/ldap_test.go:227]: Are we able to delete these commented code or do we need it?

File: api4/ldap.go [hunk 5]
```diff
@@ -290,3 +298,107 @@ func migrateIdLdap(c *Context, w http.ResponseWriter, r *http.Request) {
 	auditRec.Success()
 	ReturnStatusOK(w)
 }
+
+func parseLdapCertificateRequest(r *http.Request, maxFileSize int64) (*multipart.FileHeader, *model.AppError) {
+	err := r.ParseMultipartForm(maxFileSize)
+	if err != nil {
+		return nil, model.NewAppError("addLdapCertificate", "api.admin.add_certificate.no_file.app_error", nil, err.Error(), http.StatusBadRequest)
```
Comment (isacikgoz) [inline api4/ldap.go:305]: Would this error be anything other than having no file in the request? E.g. corrupted file or some internal errors while parsing?

File: api4/ldap_test.go (hunk 3 above)
Comment (isacikgoz) [inline api4/ldap_test.go:207]: What was the reason of commenting this code out?

File: app/ldap.go (hunk 2 above)
Comment (hahmadia) [inline app/ldap.go:255]: Needed?

File: app/ldap.go [hunk 6]
```diff
@@ -175,3 +182,101 @@ func (a *App) MigrateIdLDAP(toAttribute string) *model.AppError {
...
+}
+
+func (a *App) RemoveLdapPrivateCertificate() *model.AppError {
+	if err := a.removeLdapFile(*a.Config().LdapSettings.PrivateKeyFile); err != nil {
+		return err
+	}
+
+	cfg := a.Config().Clone()
+	*cfg.LdapSettings.PrivateKeyFile = ""
+	// *cfg.LdapSettings.Encrypt = false
```
Comment (hahmadia) [inline app/ldap.go:273]: Needed?

File: model/client4.go [hunk 7]
```diff
@@ -4079,6 +4079,50 @@ func (c *Client4) MigrateAuthToSaml(fromAuthService string, usersMap map[string]
 	return CheckStatusOK(r), BuildResponse(r)
 }
 
+// UploadLdapPublicCertificate will upload a public certificate for LDAP and set the config to use it.
+// The filename parameter is deprecated and ignored: the server will pick a hard-coded filename when writing to disk.
+func (c *Client4) UploadLdapPublicCertificate(data []byte, filename string) (bool, *Response) {
```
Comment (isacikgoz) [inline model/client4.go:4084]: Do we need to include `filename` as an arg here? Since this is not breaking the API at all.

File: i18n/en.json [hunk 8]
```diff
@@ -5374,6 +5382,10 @@
     "id": "ent.ldap.do_login.user_not_registered.app_error",
     "translation": "User not registered on AD/LDAP server."
   },
+  {
+    "id": "ent.ldap.do_login.x509.app_error",
+    "translation": ""
```
Comment (hahmadia) [inline i18n/en.json:5387]: No text required here?

**Review by isacikgoz — COMMENTED** [review]
Looks very good, have some questions below.

File: api4/ldap_test.go (hunk 3 above)
Comment (sbishel) [reply api4/ldap_test.go]: Deleted

File: app/ldap.go (hunk 2 above)
Comment (sbishel) [reply app/ldap.go]: Deleted

File: app/ldap.go (hunk 6 above)
Comment (sbishel) [reply app/ldap.go]: Deleted

**Review by isacikgoz — APPROVED** [review]
LGTM, there are some minor linter errors to be take care of.

File: api4/ldap.go [hunk 9]
```diff
@@ -32,6 +33,13 @@ func (api *API) InitLdap() {
 
 	// DELETE /api/v4/ldap/groups/:remote_id/link
 	api.BaseRoutes.LDAP.Handle(`/groups/{remote_id}/link`, api.ApiSessionRequired(unlinkLdapGroup)).Methods("DELETE")
+
+	api.BaseRoutes.LDAP.Handle("/certificate/public", api.ApiSessionRequired(addLdapPublicCertificate)).Methods("POST")
+	api.BaseRoutes.LDAP.Handle("/certificate/private", api.ApiSessionRequired(addLdapPrivateCertificate)).Methods("POST")
```
Comment (jupenur) [inline api4/ldap.go:38]: Nit: the private part is a key, not a certificate, right?

**Review by jupenur — APPROVED** [review]

**Review by srkgupta — APPROVED** [review]
Tested the feature along with its dependent PR and the Google LDAP login works fine on the webapp. Approving the PR.

**Pull request merged**